packages/apputils/src/printing.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Any object is "printable" if it implements the `IPrintable` interface.
+ *
+ * To do this it, it must have a method called `Printing.symbol` which returns either a function
+ * to print the object or null if it cannot be printed.
+ *
+ * One way of printing is to use the `printWidget` function, which creates a hidden iframe
+ * and copies the DOM nodes from your widget to that iframe and printing just that iframe.
+ *
+ * Another way to print is to use the `printURL` function, which takes a URL and prints that page.
+ */
+
+import { Widget } from '@phosphor/widgets';
+
+export namespace Printing {
+  /**
+   * Function that takes no arguments and when invoked prints out some object or null if printing is not defined.
+   */
+  export type OptionalAsyncThunk = () => Promise<void> | null;
+
+  /**
+   * Symbol to use for a method that returns a function to print an object.
+   */
+  export const symbol = Symbol('printable');
+
+  /**
+   * Objects who provide a custom way of printing themselves
+   * should implement this interface.
+   */
+  export interface IPrintable {
+    /**
+     * Returns a function to print this object or null if it cannot be printed.
+     */
+    [symbol]: () => OptionalAsyncThunk;
+  }
+
+  /**
+   * Returns whether an object implements a print method.
+   */
+  export function isPrintable(a: unknown): a is IPrintable {
+    if (typeof a !== 'object' || !a) {
+      return false;
+    }
+    return symbol in a;
+  }
+
+  /**
+   * Returns the print function for an object, or null if it does not provide a handler.
+   */
+  export function getPrintFunction(val: unknown): OptionalAsyncThunk {
+    if (isPrintable(val)) {
+      return val[symbol]();
+    }
+    return null;
+  }
+
+  /**
+   * Prints a widget by copying it's DOM node
+   * to a hidden iframe and printing that iframe.
+   */
+  export function printWidget(widget: Widget): Promise<void> {
+    return printContent(widget.node);
+  }
+
+  /**
+   * Prints a URL by loading it into an iframe.
+   *
+   * @param url URL to load into an iframe.
+   */
+  export async function printURL(url: string): Promise<void> {
+    const text = await (await fetch(url)).text();
+    return printContent(text);
+  }
+
+  /**
+   * Prints a URL or an element in an iframe and then removes the iframe after printing.
+   */
+  async function printContent(textOrEl: string | HTMLElement): Promise<void> {
+    const isText = typeof textOrEl === 'string';
+    const iframe = createIFrame();
+
+    const parent = window.document.body;
+    parent.appendChild(iframe);
+
+    if (isText) {
+      iframe.srcdoc = textOrEl as string;
+      await resolveWhenLoaded(iframe);
+    } else {
+      iframe.src = 'about:blank';
+      setIFrameNode(iframe, textOrEl as HTMLElement);
+    }
+
+    launchPrint(iframe.contentWindow);
```

Comment:
Is launching a print dialog blocking? It kind of seemed like it in my testing. No problem with that per-se, I think, but removing the iframe did seem to wait until the dialog was gone.